### ISSUE TITLE
Attendee - Checkout - Promo Code

### ISSUE THREAD

**Issue by Mcscott142** [revelrylabs/bn-mobile-react#410]
## Background 
Users should be able to apply promo codes for discounts etc on tickets. This is already implemented on the web [here](https://github.com/big-neon/bn-web/blob/5ae43696121d380f1546eb59c26871777e6085bf/src/components/pages/events/CheckoutSelection.js). 

## Scenario: A user enters a Promo Code when checking out
Given I have a promotion code
- [ ] When I enter the promo code at Tickt type selection
- [ ] And I select Apply Promo Code
- [ ] Then the discount is applied 
- [ ] And I am aware that it has been applied

## Mock 

![abstract_-_big_neon_mobile_app_-_get_ticket](https://user-images.githubusercontent.com/6877299/50108610-46ef0780-01fb-11e9-9852-eeda4659070c.png)

**Comment by jwietelmann** [comment]
Note: `redemption_code` is the cart param

**Comment by jwietelmann** [comment]
I'm not sure how we're supposed to create/find/use available valid discount codes. And sending an invalid code for `redemption_code` gets me back this, which seems like an API bug:

```
  "data": Object {
     "error": "invalid type: null, expected a UUID string at line 1 column 434",
   },
```

**Comment by zmelamed** [comment]
The UX here is still in development so it's pretty rough, but you'll go here and create a hold with a discount (in dollars) to test or a comp status to test. If you hold the entire ticket inventory for the event no tickets should be available for purchase unless the code is entered.
![big_neon_studio](https://user-images.githubusercontent.com/456953/50121066-6c384180-020c-11e9-9346-693947f9d4b8.png)

**Comment by jwietelmann** [comment]
Note to self/whoever-picks-up-next: I have created a 100-ticket, $10-discount hold for Boxing Day Brawl.

**Comment by jwietelmann** [comment]
Er, that one was already free. PICKLERICK has 100 tickets at a $10 discount for Pickle Rick's Birthday.

**Comment by Mcscott142** [comment]
Assuming @zmelamed's comment makes this unblocked, removing that label.

**Comment by daybreaker** [comment]
Ready for @brittanygay to style.

**Comment by brittanygay** [comment]
Pulling down @daybreaker's branch now to start styling!

**Comment by brittanygay** [comment]
I just commented with some updates and a screenshot here: https://github.com/revelrylabs/bn-mobile-react/pull/417#issuecomment-449459494

**Comment by brittanygay** [comment]
The styles for the promo section are now complete. Here are screenshots of the updated UI when entering a promo code and once the promo is added/shown in Checkout:

![Simulator Screen Shot - iPhone X - 2018-12-21 at 12.08.19.png](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/56154cf43bdd627e00d5004a/125516c66e82c728ace21e0d46db9d97368e82bd80fbff70a74ff5912706147b2610e366308cd320b04c7e41374f53bc43585e4ce1f88b36a4b72c21ce4c0aefcf6d576ac73403b0aac25de05bef73cf5d8fa4b9d84ac4ee6b22e3dfde4b3917501889fe0f3da09949cf58e457f30f04fa47693ebf3c633e559e0c742b616623aff9db57818166acba122e4485.png)
![Simulator Screen Shot - iPhone X - 2018-12-21 at 12.20.04.png](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/56154cf43bdd627e00d5004a/125516c66e82c728ace21e0d46db9d97368e82bd80fbff70a74ff5912706147b2610e366308cd320b04c7e41374f53bc43585e4ce1f88b36a4b72c21ce4c0aefcf6f5f6ac63903b0aac25de05bef73cf5d8fa4b9d84ac4ee6b22e3dfde4b3917501889fe0f3da09949cf58e457f30f04fa47693ebf3c633e559e0c742b616623aff9db57818166acba122d4485.png)

Passing back over to @daybreaker to finish up/look into that error mentioned in the PR. Let me know if there's anything else that may need my attention!

**Comment by brittanygay** [comment]
@daybreaker I was able to put in a fix for the promo button showing a light pink overlay when the user clicks so it's better understood that they have successfully clicked the button and the promo code is processing/brings them to the Checkout screen.

Here's a gif that shows the updates (don't pay attention to the bad gif quality of the input):
![demo-large-4.gif](https://waffleio-direct-uploads-production.s3.amazonaws.com/uploads/56154cf43bdd627e00d5004a/125516c66e82c728ace21e0d46ec91972ccf8fa89deeba0ef01fbcd6395428792c07b7692ac7b62ae9162d07585b1afd455f5611fcf98879aeb63c728d5c19f1c32f0e2a92624cf5f2d75dad0ba8218c4584b3b88f48c2ed3f2de18f884d2c53.gif)

**Comment by zmelamed** [comment]
They will still need to be able to select the number of tickets they want after they enter the promo code.

**Comment by Mcscott142** [comment]
@daybreaker the current API response doesn't allow for QTY selection right ^.

**Comment by daybreaker** [comment]
@zmelamed 	the way the bn-web site is currently working, you cant adjust quantity. 

To use a promo code, the cart update only takes a promocode with an empty items array, then automatically assigns a number of tickets equal to the ticket hold's max per order and doesnt allow you to adjust quantity.

So I just duplicated that on the mobile side.

**Comment by zmelamed** [comment]
Ok, let me pull in @Krakaw about this. 

The issue is a promo code shouldn't enforce a requirement to buy the entire max quantity allowed or the entire allocation assigned to the code... An artist may distribute a code to their friends that gives their friends access to a special allotment of tix... there may be 100 tix allotted and related to that code, and each user may be restricted to buying a max of 4, but they shouldn't be obligated to buy all 4, they should be able to buy only 1 if they want

**Comment by daybreaker** [comment]
Thats how I assumed it would work until viewing the way bn-web made the api calls.

@Krakaw 	Do I just pass in the redemption code with the normal items[] array for a ticket type and quantity? Is it only assigning `max per order` as the quantity when a redemption_code is sent with a blank items[] ?

**Comment by daybreaker** [comment]
Both options should be pretty easy to implement with how our UI is set up.

I think the way Ticket Holds are created (on a specific ticket type), option 2 will be easiest to implement in the API as well as our UI. Where adding a promo code acts just like selecting a ticket type. Then in our UI it goes to our normal quantity selection/checkout screen.

I'll play with it real quick while the deploy is going to see if maybe thats how the API already works.

**Comment by Krakaw** [comment]
Just pulling in @mikethetike for a comment on this.

**Comment by Mcscott142** [comment]
Bumping this to put back on the radar for next week. @Krakaw @mikethetike

**Comment by mikethetike** [comment]
Currently on PUT (and POST) /cart, if you add a `redemption_code`, if there are tickets in the basket already for that redemption code, the quantity can be changed. If there are not, the max per user was added. I have changed it to add the ticket type's increment.

I'm also going to add a parameter to `/events/{id}/ticket_types?redemption_code={}` to return all the ticket types including the ones via the redemption code

**Comment by mikethetike** [comment]
Chatted a bit with @Krakaw and we think the best is this:
1. PUT/POST to cart with a valid redemption code will only apply discounts or allow access to tickets that are in the cart. No new items will be added
2. We will create a new endpoint /holds/{redemption_code} to return the holds/comps/discounts from that code. This will need to be merged into the current cart manually on the frontend
 - This is different from the previous idea of reusing /events/{id}/ticket_types because there are lots of ways of retrieving ticket types and we would have to edit them all

**Comment by Mcscott142** [comment]
Just waiting on https://github.com/big-neon/bn-api/pull/862 to get merged before unblocking.

**Comment by Mcscott142** [comment]
That PR is merged and on Master. Unblocking. cc @daybreaker

**Comment by Mcscott142** [comment]
Looks good! Moving to UAT.

**Comment by Mcscott142** [comment]
Checked stage build via TestFlight/Google Play and looks good to me.
